tests/steamship_tests/app/unit/test_mixins.py

```diff
@@ -0,0 +1,45 @@
+import pytest
+from assets.packages.package_with_mixins import PackageWithMixin, TestMixin
+
+from steamship import SteamshipError
+from steamship.invocable import Invocable, InvocableRequest, Invocation
+from steamship.utils.url import Verb
+
+
+def invoke(o: Invocable, path: str):
+    req = InvocableRequest(invocation=Invocation(http_verb="POST", invocation_path=path))
+    return o(req).data
+
+
+def test_package_with_mixin_routes():
+    """Tests that we can inspect the package and mixin routes"""
+    package = PackageWithMixin()
+    assert (
+        package._package_spec.method_mappings[Verb.POST]["/test_mixin_route"].func_binding
+        is not None
+    )
+    assert (
+        package._package_spec.method_mappings[Verb.POST]["/test_package_route"].func_name_binding
+        == "test_package_route"
+    )
+    assert invoke(package, "test_mixin_route") == "mixin yo"
+    assert invoke(package, "test_package_route") == "package"
+
+    routes = [m["path"] for m in package.__steamship_dir__()["methods"]]
+    assert "/test_mixin_route" in routes
+    assert "/test_package_route" in routes
+    assert len(routes) == 5  # __instance__init + 2x __dir__
+
+
+def test_instance_init():
+    """Tests that instance init is called on the mixins"""
+    package = PackageWithMixin()
+    package.invocable_instance_init()
+
+    assert package.mixins[0].instance_init_called
+
+
+def test_two_conflicting_mixins_raises_error():
+    package = PackageWithMixin()
+    with pytest.raises(SteamshipError):
+        package.add_mixin(TestMixin("ahoy"))
```

Comment:
How do you feel about plumbing the `permit_overwrite_of_existing` variable all the way up to `add_mixin` but defaulting it to `False`? I don't think I feel strongly, but seemed like an obvious question to ask

Reply:
I'm not sure how it would work.  The mixin will have its own internal state, which may be affected by the calling of its routes.  It assumes a pairing of that internal state and which routes it defines.  Permitting overwrite means being allowed to be overwritten, in which case bad things almost certainly start happening, right?

If you want to extend or change the behavior of a mixin, I think you'd want to subclass the mixin, and then mix THAT in :)

Reply:
Yeah I think there's def enough murkiness there for the default to be: "you can't clobber unless you know what you're doing so much you can figure out how to clobber manually" :)